docs/unittesting.md

```diff
@@ -0,0 +1,264 @@
+
+```
+...
+"Logged errors:"
+:demo/results/2020.11.11/logs/err_eod.log| ()
+:demo/results/2020.11.11/logs/err_wdb.log| ,"2020.11.11D16:37:21.716171000|aquaq-184|test|test1|ERR|KUexecerr|run error in file :/ho.."
+...
+q) raze value errors
+"2020.11.11D16:37:21.716171000|aquaq-184|test|test1|ERR|KUexecerr|run error in file :/home/mpotter/kdbCode/segtp/deploy/tests/demo/wdb/test.csv on line 8 - stHandle. Code: 'stHandle @/: `.u.upd ,/: ((`trade;testtrade);(`quote;testquote))'"
+```
+
+The script has read the file `demo/results/2020.11.11/logs/err_wdb.log` and from the message within we can see that there is a code error in a 'run' command in the file `/home/mpotter/kdbCode/segtp/deploy/tests/demo/wdb/test.csv` on line 8, and the offending piece of code is also displayed. This may be enough information for us to solve the issue, but if not we can dig deeper. 
+
+We can see that the error is coming from the WDB tests, and so we run those in debug mode in the following way:
+
+```shell
+./demo/wdb/run.sh -d
+```
+
+This runs the tests and displays the error message we saw earlier before outputting our test results and failures and leaving us in the q session. By examining the error message and being able to access the q session we can see that the variable `stHandle` should in fact be `stpHandle` and this typo is causing the error. Once this is fixed, it can be run again and the error doesn't appear any more. There is still one test failing, however:
+
+```
+q) select action,code,csvline from KUerr
+action code                                               csvline
+-----------------------------------------------------------------
+true   (t1+1 5)~t2:wdbHandles[`all`sym] @\: "count trade" 10
+```
+
+The best way to debug this would be to be able to exit the tests as this line is being run and be able to examine it from there. We can do this by invoking the stop mode in our test script:
+
+```shell
+./demo/wdb/run.sh -s
+```
+
+This throws us out to the q prompt at this point in the tests with the following error:
+
+```
+'failed to load /home/mpotter/kdbCode/segtp/deploy/tests/runtests.q : true test failure in file :/home/mpotter/kdbCode/segtp/deploy/tests/demo/wdb/test.csv on line 10
+```
+
+We can get the code which failed and run it here to see what it returns. From doing some quick debugging we can see that one of the items being added to `t1` is wrong, it should be ten rather than 1. Once this is fixed we can run the WDB tests again and we see that there are now no errors and all the tests pass! We can then run all of our tests again as at the start and no new test failures come up and the latest error logs are empty.
+
+### Notes on Best Practice
+
+Here are some recommendations on making test development more straightforward:
+
+- Use TorQ connection management when dealing with multiple processes
+- Run the tests from a 'blank' test process where possible, as this ensures that test code and process code don't get in each other's way
```

Comment:
I think I disagree with this point, related to my comment above; in general, the whole point is to test "process code", so loading a blank test process complicates things. Typically better IMO to load a test process with relevant `-proctype` so process config & code can be loaded to be tested

Reply:
I see what you mean, I guess I'm thinking about this mainly from an integration testing perspective where you'll have various processes interacting, whereas from a unit testing perspective your approach makes more sense. I'll try and make the distinction clearer.